Factorize cmake logic between core and algorithms perf tests

tretre91 opened the pull request:
This PR factorizes the cmake logic between the core and algorithms perf tests, we avoid duplicating the code for finding google benchmark and only define one `kokkos_add_benchmark` macro which can be reused by the other perf tests.

The `kokkos_add_benchmark` macro now takes an optional argument `MAIN_FILE` which allows using a specific BenchmarkMain.cpp file instead of the one in core/perf_test, this is needed for https://github.com/kokkos/kokkos/pull/7932

tretre91 committed:
Find googlebench and define macros only in the core perf test CMakeLists

Signed-off-by: Trévis Morvany <trevis.morvany@cea.fr>

tretre91 committed:
Cleanup
- Use the `kokkos_add_benchmark_directories` macro instead of adding a
  new `kokkos_add_bencmark_directory` function
- Remove some useless checks for `KOKKOS_ENABLE_BENCHMARKS`

Signed-off-by: Trévis Morvany <trevis.morvany@cea.fr>

tretre91 committed:
Add core/perf_test as include_dir for BenchmarkMain.hpp

Signed-off-by: Trévis Morvany <trevis.morvany@cea.fr>

dalg24 commented on core/perf_test/CMakeLists.txt:
```diff
@@ -49,17 +26,24 @@ else()
 
-function(KOKKOS_ADD_BENCHMARK NAME)
-  cmake_parse_arguments(BENCHMARK "" "" "SOURCES" ${ARGN})
+macro(KOKKOS_ADD_BENCHMARK NAME)
```
Why change it to a macro?

tretre91 replied:
So that we don't have to specify an absolute path or prefix the source files with `${CMAKE_CURRENT_LIST_DIR}` when using it from other perf test directories

dalg24 replied:
Maybe add a comment above the macro definition to capture this

tretre91 committed:
Add a comment explaining why we use a macro for kokkos_add_benchmark

Signed-off-by: Trévis Morvany <trevis.morvany@cea.fr>

tretre91 committed:
Add GLOBAL keyword to find_package(benchmark)
- without this the imported target is not visible from other directories
  when we use an already installed google benchmark

Signed-off-by: Trévis Morvany <trevis.morvany@cea.fr>

tretre91 committed:
Call find_package in the root cmake file instead of using GLOBAL
- GLOBAL is only available for cmake >=3.24

Signed-off-by: Trévis Morvany <trevis.morvany@cea.fr>

tretre91 commented:
This wasn't working with an already installed google benchmark picked up by `find_package`, as the alias target `google::benchmark` was only visible in `core/perf_test`. There is a `GLOBAL` keyword for `find_package` which makes the target visible at global scpoe in the project, but it's in cmake >=3.24, so I just moved the call to `find_package(benchmark ...)` to the root CMakeLists.txt

masterleinad commented on core/perf_test/CMakeLists.txt:
```diff
@@ -49,17 +25,26 @@ else()
   set(BENCHMARK_NAME Kokkos_${NAME})
-  list(APPEND BENCHMARK_SOURCES BenchmarkMain.cpp Benchmark_Context.cpp)
+  set(CORE_PERFTEST_DIR ${PROJECT_SOURCE_DIR}/core/perf_test)
+  if(BENCHMARK_MAIN_FILE)
+    set(BENCHMARK_MAIN ${BENCHMARK_MAIN_FILE})
```
Where is this used?

tretre91 replied:
This will be used in https://github.com/kokkos/kokkos/pull/7932, but for the moment this is unused

masterleinad reviewed (approved):
Fine with me.
